### PR TITLE
Add transaction index and timestamps to events

### PR DESCRIPTION
**Pull request by jimthematrix**
This helps with sequencing on the client side when there are multiple events fired in the same block.

The change also replaces the original block parsing logic that used `filtered transactions` from the Fabric library, with manually parsing the entire block using raw proto definitions. This allows us to get richer information out of the block.

Also added the timestamp (unix nano) to events:
- for subscriptions listening on blocks (no `chaincodeId` specified), timestamp is added from the timestamp parsed out of the block
- for chaincode events, since the event itself pushed by Fabric doesn't contain the timestamp, we must make a call to Fabric using the event's block number. This call is optional depending on the `timestamps` property in the stream specification

### CODE REVIEW

File: internal/events/subscription.go [hunk 1]
```diff
@@ -140,13 +140,35 @@ func (s *subscription) processNewEvents() {
...
 		}
 	}
 }
 
+func (s *subscription) getEventTimestamp(evt *eventsapi.EventEntry) {
+	// the key in the cache is the block number represented as a string
+	blockNumber := strconv.FormatUint(evt.BlockNumber, 10)
+	if ts, ok := s.ep.stream.blockTimestampCache.Get(blockNumber); ok {
+		// we found the timestamp for the block in our local cache, assert it's type and return, no need to query the chain
+		evt.Timestamp = ts.(int64)
```
Comment (nguyer) [inline internal/events/subscription.go:158]: Should we check the result of this casting?

**Review by nguyer — APPROVED** [review]
Looks good. Minor question on some error checking.

File: internal/events/subscription.go (hunk 1 above)
Comment (peterbroadhurst) [reply internal/events/subscription.go:158]: I think this is ok because the only code that adds to this cache is in this same function

**Review by peterbroadhurst — APPROVED** [review]

**Comment by peterbroadhurst** [comment]
This is a critical piece to merging the `onchain-logic` code to main, so have also done a review to merge.

**Pull request merged**